modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/CriteriaBuilder.es.js

```diff
@@ -0,0 +1,315 @@
+						]
+					)
+				)
+			}
+		),
+		modelLabel: PropTypes.string,
+		onChange: PropTypes.func,
+		supportedConjunctions: PropTypes.arrayOf(
+			PropTypes.shape(
+				{
+					label: PropTypes.string,
+					name: PropTypes.string.isRequired
+				}
+			)
+		),
+		supportedOperators: PropTypes.array,
+		supportedProperties: PropTypes.arrayOf(
+			PropTypes.shape(
+				{
+					entityUrl: PropTypes.string,
+					label: PropTypes.string,
+					name: PropTypes.string.isRequired,
+					options: PropTypes.array,
+					type: PropTypes.string.isRequired
+				}
+			)
+		).isRequired,
+		supportedPropertyTypes: PropTypes.object
+	};
+
+	static defaultProps = {
+		readOnly: false
+	};
+
+	state = {
+		editing: false
+	};
+
+	/**
+	 * Cleans the criteria by performing the following:
+	 * 1. Remove any groups with no items.
+	 * 2. Flatten groups that directly contain a single group.
+	 * 3. Flatten groups that contain a single criterion.
+	 * @param {array} criteriaItems A list of criterion and criteria groups
+	 * @param {boolean} root True if the criteriaItems are from the root group.
+	 * to clean.
+	 */
+	_cleanCriteriaMapItems(criteriaItems, root) {
+		return criteriaItems
+			.filter(
+				({items}) => {
+					return items ? items.length : true;
+				}
+			)
+			.map(
+				item => {
+					let cleanedItem = item;
+
+					if (item.items) {
+						if (item.items.length === 1) {
+							const soloItem = item.items[0];
+
+							if (soloItem.items) {
+								cleanedItem = {
+									conjunctionName: soloItem.conjunctionName,
+									groupId: soloItem.groupId,
+									items: this._cleanCriteriaMapItems(soloItem.items)
+								};
+							}
+							else {
+								cleanedItem = root ? item : soloItem;
+							}
+						}
+						else {
+							cleanedItem = {
+								...item,
+								...{
+									items: this._cleanCriteriaMapItems(item.items)
+								}
+							};
+						}
+					}
+
+					return cleanedItem;
+				}
+			);
+	}
+
+	/**
+	 * Switches the edit state between true and false.
+	 */
+	_handleToggleEdit = () => {
+		this.setState(
+			{
+				editing: !this.state.editing
+			}
+		);
+	};
+
+	/**
+	 * Cleans and updates the criteria with the newer criteria.
+	 * @param {Object} newCriteria The criteria with the most recent changes.
+	 */
+	_handleCriteriaChange = newCriteria => {
+		this.props.onChange(this._cleanCriteriaMapItems([newCriteria], true).pop());
+	};
+
+	/**
+	 * Moves the criterion to the specified index by removing and adding, and
+	 * updates the criteria.
+	 * @param {string} startGroupId Group ID of the item to remove.
+	 * @param {number} startIndex Index in the group to remove.
+	 * @param {string} destGroupId Group ID of the item to add.
+	 * @param {number} destIndex Index in the group where the criterion will
+	 * be added.
+	 * @param {object} criterion The criterion that is being moved.
+	 * @param {boolean} replace True if the destIndex should replace rather than
+	 * insert.
+	 */
+	_handleCriterionMove = (
```

Comment:
Since all of these arguments are just copied from `_searchAndUpdateCriteria ` do you think it'd make more sense to do something like this?

```js
_handleCriterionMove = (...args) => {
		const newCriteria = this._searchAndUpdateCriteria(
			this.props.criteria,
			...args
		);

 		this._handleCriteriaChange(newCriteria);
	}
```

Reply:
I think that can work since there's documentation on the method anyway.